src/core/utils/request/urlFromRequestOrCache.ts

```diff
@@ -0,0 +1,17 @@
+const urlCache = new Map<string, URL>()
+
+/**
+ * Returns the URL object for the given request. The URL object is cached so that
+ * subsequent calls to this function with the same request or requests to the same
+ * url will return the same URL object, to avoid multiple URL object allocations and
+ * parsing the same url multiple times.
+ */
+export function urlFromRequestOrCache(request: Request): URL {
+  const requestUrl = request.url
+  let url = urlCache.get(requestUrl)
+  if (!url) {
+    url = new URL(requestUrl)
```

Comment:
URL allocations aren't cheap as they do expensive string parsing to create a URL object.  we can cache this per string url, and avoid those object allocations and parses, which may occur multiple times for a single request lookup

Reply:
Okay, I see. Thanks for a brief explanation!

Reply:
They're also not super expensive, so definitely a micro-optimization we could totally skip 

It's definitely the least important change.  on the order of 10ms for 10,000 requests of overhead.  doubtful that it'll matter to most!

<img width="1037" alt="Screen Shot 2024-01-07 at 10 57 55 PM" src="https://github.com/mswjs/msw/assets/8616962/d28133e8-fa57-4069-8e49-732bb0fc3f02">

<img width="1035" alt="Screen Shot 2024-01-07 at 10 53 12 PM" src="https://github.com/mswjs/msw/assets/8616962/7b4a0eaa-4e43-4f5b-a161-4628cc443967">

Reply:
I'd say 10ms save on 10,000 requests is not worth the price we are paying here. Would you agree that we can skip this pull request as of now?